drivers/sdhc/sdhci_common.c

```diff
@@ -0,0 +1,801 @@
+
+	hdc_cmd.opcode = SD_STOP_TRANSMISSION;
+	hdc_cmd.arg = 0;
+	hdc_cmd.response_type = SD_RSP_TYPE_R1;
+	hdc_cmd.timeout_ms = 1000;
+
+	return sdhci_send_cmd(sdhci_ctx, &hdc_cmd, false);
+}
+
+static bool sdhc_need_cmd12(struct sdhc_command *cmd)
+{
+	return (cmd->opcode == SD_READ_MULTIPLE_BLOCK || cmd->opcode == SD_WRITE_MULTIPLE_BLOCK);
+}
+
+static int sdhci_poll_data_complete(struct sdhci_common *sdhci_ctx, int32_t timeout)
+{
+	uint32_t reg_normal_int_stat = 0;
+	int ret = -ETIMEDOUT;
+
+	timeout = timeout ? timeout : 1;
+	timeout = timeout * 100;
+
+	while (timeout > 0) {
+		reg_normal_int_stat = sys_read32(sdhci_ctx->reg_base + SDHCI_INT_STATUS);
+		if (reg_normal_int_stat & SDHCI_INT_XFER_COMPLETE) {
+			sys_write32(SDHCI_INT_XFER_COMPLETE,
+				    sdhci_ctx->reg_base + SDHCI_INT_STATUS);
+			ret = 0;
+			break;
+		}
+
+		k_busy_wait(10);
+		timeout--;
+	}
+
+	if (reg_normal_int_stat & SDHCI_INT_ERROR) {
+		LOG_ERR("sdhc:req: data xfer error int_status:%08x", reg_normal_int_stat);
+		sys_write32(SDHCI_INT_ERROR_MASK, sdhci_ctx->reg_base + SDHCI_INT_STATUS);
+		ret = -EIO;
+	}
+
+	if (!timeout) {
+		LOG_ERR("sdhc:req: data xfer timeout int_status:%08x", reg_normal_int_stat);
+	}
+
+	return ret;
+}
+
+static void sdhci_data_port_io(struct sdhci_common *sdhci_ctx, uint8_t *data, uint32_t size,
+			       bool read)
+{
+	uint32_t *offs = (uint32_t *)data;
+	int i;
+
+	for (i = 0; i < size; i += 4, offs++) {
+		if (read) {
+			*offs = sys_read32(sdhci_ctx->reg_base + SDHCI_BUFFER);
+		} else {
+			sys_write32(*offs, sdhci_ctx->reg_base + SDHCI_BUFFER);
+		}
+	}
+}
+
+static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data *data,
+			       bool is_read)
+{
+	uint32_t block = data->blocks;
+	uint32_t reg_present_state;
+	uint32_t int_status, rdy_mask;
+	uint8_t *data_ptr = data->data;
+	int32_t timeout;
+
+	timeout = data->timeout_ms ? data->timeout_ms : 1;
+	timeout = timeout * 100;
+
+	if (is_read) {
+		rdy_mask = SDHCI_BUF_RD_ENABLE;
+	} else {
+		rdy_mask = SDHCI_BUF_WR_ENABLE;
+	}
+
+	while (timeout && block) {
+		int_status = sys_read32(sdhci_ctx->reg_base + SDHCI_INT_STATUS);
+		if (int_status & SDHCI_INT_ERROR) {
+			LOG_ERR("sdhc:req: data xfer error int_status:%08x", int_status);
+			return -EIO;
+		}
+
+		reg_present_state = sys_read32(sdhci_ctx->reg_base + SDHCI_PRESENT_STATE);
+		if (reg_present_state & rdy_mask) {
+			sdhci_data_port_io(sdhci_ctx, data_ptr, data->block_size, is_read);
+			data_ptr += data->block_size;
+			block--;
+		} else {
+			k_busy_wait(10);
+			timeout--;
+		}
+	};
+
+	if (!timeout) {
+		LOG_ERR("sdhc:req: transfer data timeout");
+		return -ETIMEDOUT;
+	}
+
+	return sdhci_poll_data_complete(sdhci_ctx, timeout / 100);
+}
+
+static void sdhci_prepare_dma(struct sdhci_common *sdhci_ctx, struct sdhc_data *data,
+			      bool is_read)
+{
+	uintptr_t dma_addr;
+
+	if (!is_read) {
+		sys_cache_data_flush_range(data->data, (data->blocks * data->block_size));
```

Comment:
You need to flush caches in both cases: read and write. Consider the following sequence for reading:

1. Client allocates read buffer
2. Client makes `memzero(read_buffer)` -> this gets cached.
3. Client starts SDHC read operation to `read_buffer`
4. HW writes to `read_buffer`
4. **At the same time** CPU decides to flush caches and zeroes from p.2 overwrite newly read data.

Reply:
Huh. I can update it this way - np. But that's how most of Zephyr drivers doing it :(

Reply:
updated